Add Yul semantics for `CALL*` operations

Coda-Coda opened the pull request:
This also includes the widespread changes necessary to have `ExecutionEnv.code` be either EVM bytecode or Yul code.

Julek committed:
Factored in EVM stack operations into EVMYul.step function

Coda-Coda committed:
Make `ExecutionEnv` depend on EVM/Yul `OperationType`

In preparation for handling CALL* operations in the Yul semantics

Coda-Coda committed:
Make `PersistentAccountState` depend on EVM/Yul `OperationType`

In preparation for handling CALL* operations in the Yul semantics

Coda-Coda committed:
Fix flow on effects of using EVM/Yul `OperationType`

Includes changes to `step` in `Semantics.lean` that removes previously defined Yul semantics from operations where this no longer is applicable.

Coda-Coda committed:
Add `BEq` and `Repr` instances as some are needed for the conformance tests

This commit was made with the assistance of Cursor (cursor.com).

Coda-Coda committed:
Handle EVM/Yul `OperationType` in conformance tests

Coda-Coda committed:
Move `.Yul, _` case to the end of the match statement

Coda-Coda committed:
Add semantics for POP for Yul

As discussed.

Coda-Coda committed:
Finmap approach to modelling external calls

Termination proof not complete.

Coda-Coda committed:
Simplify and fix termination proofs by always reducing fuel

Coda-Coda committed:
Create `YulContract` structure with an initial model of a dispatcher

Termination proof still broken.

Coda-Coda committed:
Remove `ExternallCall` branch from `Stmt`

This is the wrong approach.

Coda-Coda committed:
Change `dispatcher` to `Yul.Ast.Stmt`

Coda-Coda committed:
Change to `YulFunctionName/String` for `Expr` and `Stmt`

In preparation for calling functions via their name rather than through and inlined `Yul.Ast.Stmt`.

Coda-Coda committed:
Make Yul semantics call functions using their name (as a String)

Coda-Coda committed:
Combine `PrimCall` and `Call` in `Ast.lean`

Coda-Coda committed:
Fix Yul examples based on new `Stmt` and `Expr` structure

Coda-Coda committed:
Fix `SizeLemmas` based on new `Stmt` and `Expr` structure

Coda-Coda committed:
Combine `Call` and `PrimCall` in Yul semantics

And fix Yul semantics from changes to `Stmt` and `Expr` structure

Coda-Coda committed:
Clearly give `.Assign` the same semantics as `.Let`

Similarly for `.AssignCall`.

Coda-Coda committed:
Undo combining `PrimCall` and `Call` for `ExprStmt`

To better take into account: "Thus, we cannot have literals or variables on the RHS."

Coda-Coda committed:
Combine cases further in `Stmt`

And adjust notation examples and the Yul semantics as required by the change.

Coda-Coda committed:
Rename `inputData` -> `calldata`

Coda-Coda committed:
Add Yul semantics for `call`

Coda-Coda committed:
Handle writing return data to memory as specified

See: https://docs.soliditylang.org/en/latest/yul.html#yul-call-return-area

Coda-Coda committed:
Handle parsing of Yul function calls

Due to calling functions using their name (as a string) instead of inlining the `FunctionDefinition`.

Coda-Coda committed:
Generalise `Operation.pretty` and add debug logging for Yul ops

Coda-Coda committed:
Add newline to highlight invocation of `call`

Coda-Coda committed:
Improve semantics of `CALL`

Handling the slightly less clear cases based on the yellow paper, comparing with the EVM semantics in this repo, and considering in particular the Nethermind execution client.

Coda-Coda committed:
Add `callFromCode` in preparation for adding `callcode` semantics

Coda-Coda committed:
Fix handling of memory in CALL Yul semantics

Coda-Coda committed:
Add Yul semantics for `STATICCALL`

Coda-Coda committed:
Add Yul semantics for `CALLCODE`

Coda-Coda committed:
Fix overflow below zero handling in transfers

Coda-Coda committed:
Add Yul semantics for `DELEGATECALL`

Coda-Coda committed:
Fix callvalue transfer being lost

Coda-Coda committed:
Increase max heartbeats

Coda-Coda committed:
Replace FFI `ByteArray.zeroes` with a native Lean definition

Coda-Coda committed:
Switch to using `Except` for Yul semantics

Coda-Coda committed:
Revert "Replace FFI `ByteArray.zeroes` with a native Lean definition"

This reverts commit 53f2460998c823d77e4f70afa7ad1383763fbca5.

Coda-Coda committed:
Replace returning `default` to indicate an error with raising relevant `Yul.Exception`s

And add `execTopLevel` which describes how different exceptions should be interpreted. `default` there indicates an error that should not be possible if the state is set up correctly and the compiler runs correctly.

Coda-Coda committed:
Add Repr instance for Yul exceptions

Coda-Coda committed:
Add initial test of a cross-contract call

Coda-Coda committed:
Model Yul halting for `RETURN` via a special Exception

Coda-Coda committed:
Restore `ExecutionEnv` after an external CALL

Coda-Coda committed:
Remove `value` arg and handling from `STATICCALL`

The value is implicitly zero, because it is a *static* call.

Coda-Coda committed:
Add missing import

Coda-Coda committed:
Add Yul Exceptions

Coda-Coda committed:
Make `EXTCODESIZE` return an error for Yul

To make it clearer that this operation is undefined for Yul.

Coda-Coda committed:
Tweak generated Yul to not use `extcodesize`

And assume that the target address is a smart contract.

Coda-Coda committed:
Apply the transfer in the case where the contract does not exist

Like the STOP opcode.

This commit was made with the assistance of Cursor (cursor.com) using gpt-5.

Coda-Coda committed:
Also set `returnData` to `empty`

This commit was made with the assistance of Cursor (cursor.com) using gpt-5.

Coda-Coda committed:
Fix destination for copying `H_return`

This commit was made with the assistance of Cursor (cursor.com) using gpt-5.

Coda-Coda committed:
Ensure `memory` is reset before external contract calls

This commit was made with the assistance of Cursor (cursor.com) with use of gpt-5.

Coda-Coda committed:
Set `H_return` to zero array at the end of a successful call.

This commit was made with the assistance of Cursor (cursor.com).

Coda-Coda committed:
Handle Yul halting in `STATICCALL`/`DELEGATECALL`/`CALLCODE`

Coda-Coda committed:
Fix spelling

And related function selector

andreiburdusa commented on EvmYul/EVM/Semantics.lean:
```diff
@@ -232,69 +232,8 @@ def step (fuel : ℕ) (gasCost : ℕ) (instr : Option (Operation .EVM × Option
         | .none => fetchInstr evmState.toState.executionEnv evmState.pc
         | .some (instr, arg) => pure (instr, arg)
     let evmState := { evmState with execLength := evmState.execLength + 1 }
+    -- EVLYul.step instr {evmState with { gasAvailable := if tracking_gas then evmState.gasAvailable - UInt256.ofNat gasCost else evmState.gasAvailable}}
```
Should this comment be removed?

Coda-Coda replied:
Yes, will do.

andreiburdusa commented on EvmYul/Yul/Ast.lean:
```diff
@@ -30,37 +30,68 @@ instance : ToString PrimOp := ⟨stringOfPrimOp⟩
+    | OperationType.Yul => YulContract
+
+instance {τ} : BEq (contractCode τ) where
+  beq a b := (match τ with
```
Parentheses are not needed in the rhs

Coda-Coda replied:
Will do.

andreiburdusa commented on EvmYul/Yul/Ast.lean:
```diff
@@ -30,37 +30,68 @@ instance : ToString PrimOp := ⟨stringOfPrimOp⟩
+              | .Yul => a == b)
+
+instance {τ} : Inhabited (contractCode τ) where
+  default := (match τ with
```
Parentheses are not needed in the rhs

andreiburdusa commented on EvmYul/Yul/Ast.lean:
```diff
@@ -30,37 +30,68 @@ instance : ToString PrimOp := ⟨stringOfPrimOp⟩
+                | .Yul => default)
+              
+instance {τ} : Repr (contractCode τ) where
+  reprPrec a _ := (match τ with
```
Parentheses are not needed in the rhs

andreiburdusa commented on EvmYul/Yul/YulSemanticsTests/Main.lean:
```diff
@@ -0,0 +1,526 @@
+open EvmYul.Yul
+
+def main : IO Unit := do
+  IO.println s!"test₁: {test₁}"
```
I can't `#eval main` because we don't have a native implementation of `ffi.ByteArray.zeroes`. Not sure what should be done about this... remove `test₁` altogether or come up with a better one :thinking:

Coda-Coda replied:
This test must be run via `lake exe yulSemanticsTests`.
I will add a comment just before `main` to indicate this. This way we can run the foreign C function. I considered replacing `ByteArray.zeroes` with a native definition [here](https://github.com/NethermindEth/EVMYulLean/pull/79/commits/53f2460998c823d77e4f70afa7ad1383763fbca5) but decided to revert it ([here](https://github.com/NethermindEth/EVMYulLean/pull/79/commits/6dc89eface8375483ed74400dc5a75ffb93d4a0c)) after the performance tests failed after that change (however, the performance tests still fail for me, so perhaps the native implementation is okay).

andreiburdusa commented on EvmYul/Yul/Interpreter.lean:
```diff
@@ -19,206 +21,697 @@ open Ast SizeLemmas
 
+mutual
 /--
 TODO: Temporary EvmYul artefact with separate primop implementations.
```
Is `primCall` temporary anymore? :)

Coda-Coda replied:
👍 Will remove the solved TODO comment.

andreiburdusa commented on EvmYul/Yul/Interpreter.lean:
```diff
@@ -19,206 +21,697 @@ open Ast SizeLemmas
+            match yulFunctionNameOption with
+              | .none => .some (FunctionDefinition.Def [] [] [yulContract.code.dispatcher])
+              | .some yulFunctionName =>
+                  yulContract.code.functions.lookup yulFunctionName
```
This part is not clear to me. I expected "dispatcher" to be name of the mechanisms that links the function name to its implementation, if the implementation exists.

Coda-Coda replied:
`solc` generates for us Yul code for the dispatcher as a specially marked function with no name. This is then wrapped in `<s ... >` manually or by a script and incorporated into the state after `dispatcher := ...` e.g. see [here](https://github.com/NethermindEth/EVMYulLean/blob/6c460ddce382483717ca08da94dbf4a74ac9dd5f/EvmYul/Yul/YulSemanticsTests.lean#L20).

I will add some documentation related to this.

When `.none` is passed as the `yulFunctionNameOption` we treat this as the dispatcher being called. However, as you point out in another comment, we always pass `.none` and `[]` to `callFromCode` so this is somewhat redundant as every time a cross-contract call happens it is routed to the dispatcher. I will simplify this.

Coda-Coda replied:
Documentation added [here](https://github.com/NethermindEth/EVMYulLean/blob/4fa832ee315f4ae6f51f92b34fd6536344f8ed5a/EvmYul/Yul/YulSemanticsTests/README.md).

andreiburdusa commented on EvmYul/Yul/Interpreter.lean:
```diff
@@ -19,206 +21,697 @@ open Ast SizeLemmas
+            match yulFunctionNameOption with
+              | .none => .some (FunctionDefinition.Def [] [] [s.executionEnv.code.dispatcher])
+              | .some yulFunctionName =>
+                  s.executionEnv.code.functions.lookup yulFunctionName
```
As in the case of `call`, this part is not clear to me. I'm also wondering why the parameters `yulFunctionNameOption` and `args` are needed in the first place since they anre `.none` and `[]` in every call site of `callFromCode`.

Coda-Coda replied:
You are right, this is redundant. I will simplify this.
This happened because earlier in the development of it, I thought it might be possible to call an external contract's (non-dispatcher) function directly, but this is not the case. All external calls are routed through the dispatcher.

andreiburdusa commented on EvmYul/Yul/Interpreter.lean:
```diff
@@ -19,206 +21,697 @@ open Ast SizeLemmas
+        match args with
+          | _ :: address_arg :: value :: inOffset :: inSize :: outOffset :: outSize :: _ =>
+            if ¬s₀.executionEnv.perm ∧ value ≠ ⟨0⟩
+            then .error .StaticModeViolation
```
If you add a `do` at the beginning of the implementation of `primCall` you can replace this with
```if ¬s₀.executionEnv.perm ∧ value ≠ ⟨0⟩ then throw .StaticModeViolation```
and save some indentation. There might be some other place where you can apply this trick as well

Coda-Coda replied:
Thanks for the tip! I've done this for each time `.StaticModeViolation` is raised in `Interpreter.lean`.

andreiburdusa commented on EvmYul/Yul/Interpreter.lean:
```diff
@@ -19,206 +21,697 @@ open Ast SizeLemmas
+                    | .Ok sharedState₀ varstore =>
+                      let sharedState₁ := {sharedState₀ with H_return := ByteArray.empty,
+                                                             returnData := ByteArray.empty }
+                      .ok (.Ok sharedState₁ varstore, [⟨0⟩]) -- Insufficient funds: return 0 to indicate error, with empty return data 
```
This pattern seems to repeat quite a lot. Maybe factoring it out would make `primCall` shorter and easier to follow?

Coda-Coda replied:
Yes, good point, will do.

Coda-Coda committed:
Remove unnecessary comment

Coda-Coda committed:
Remove unnecessary parentheses

Coda-Coda committed:
Add comment to indicate how to run Yul semantics tests

Coda-Coda committed:
Delete resolved TODO

Coda-Coda committed:
Simplify `callFromCode` and rename to `callDispatcher`

Coda-Coda committed:
Use do notation to reduce indentation

Coda-Coda committed:
Refactor

Create helper function `buildContractCallEmptyReturnState`.

Coda-Coda committed:
Remove repetitition of comments

Coda-Coda committed:
Move and rename `YulSemanticsTests.lean`

Coda-Coda committed:
Add Yul semantics test instructions

Coda-Coda committed:
Add source code for contracts for Yul semantics tests

Coda-Coda committed:
Clarify syntax

Coda-Coda committed:
Add limitations of the Yul semantics to README

Coda-Coda committed:
Fix check for `perm` at start of `STATICCALL`

Coda-Coda committed:
Restore `ExecutionEnv` for `STATICCALL`/`DELEGATECALL`/'CALLCODE`

Coda-Coda committed:
Add `caller2` example Yul contract

Coda-Coda committed:
Add test of a succesful static call

Coda-Coda committed:
Set the Storage contract to have an initial `number` of 21 stored

Coda-Coda committed:
Add output of success or failure for existing Yul tests

Coda-Coda committed:
Remove debug logging

Coda-Coda committed:
Amend `Caller2` test contract to pass a value to store

This should result in a `.StaticModeViolation` when run.

This commit was made with the assistance of Cursor (cursor.com).

Coda-Coda committed:
Simplify accessing `executionEnv`

Coda-Coda committed:
Use state that has had `perm` set to `false`

Coda-Coda committed:
Fix error propogation during switches

Before this fix, an error (such as `.STATICMODEVIOLATION`) in any of the branches would cause the final result to be the error (except for `.YulHalt`). The correct behaviour is for only that branch to be associated with the error, and not the whole result of the switch. This is now the behaviour.

Coda-Coda committed:
Add check for primOps that violate static mode

Coda-Coda committed:
Add test for `fun_testStaticStore`

Coda-Coda committed:
Add a test of `delegateCall`

Coda-Coda committed:
Fix implementation of `DELEGATECALL`

The previous implementation only used the other contract's context for the call to the dispatcher.

Coda-Coda committed:
FIx `CALLCODE` semantics

It should use the code from `address` but otherwise stay in the context of the current contract.

Coda-Coda committed:
Add test for `CALLCODE`

The Yul code has been manually altered from `pop(delegatecall(_3,  4,  add(expr_mpos,  0x20),  _2, 0, 0))` to `pop(callcode(_3,  4, 0, add(expr_mpos,  0x20),  _2, 0, 0))` (adding the value argument, zero, and changing delegate to call. Solidity no longer outputs `callcode` except when using inline assembly, preferring `delegatecall`.

Coda-Coda committed:
Add `callcode` to parser

Coda-Coda committed:
Update Solidity and Yul source files for Yul semantics testing

Coda-Coda committed:
Improve README

andreiburdusa reviewed (approved):
Looks good to me

Coda-Coda committed:
Add Yul halting for `REVERT` and `STOP`

Coda-Coda committed:
Add info about SELFDESTRUCT limitations to README

Coda-Coda commented on EvmYul/Semantics.lean:
```diff
@@ -215,18 +215,18 @@ def swap (n : ℕ) : Transformer .EVM :=
       match τ with
         | .EVM => λ evmState ↦ .ok <| {evmState with toMachineState := evmState.toMachineState.setReturnData .empty}
-        | .Yul => λ yulState _ ↦ .ok (yulState, none)
+        | .Yul => λ yulState _ ↦ .error (Yul.Exception.YulHalt yulState ⟨0⟩)
```
I was looking at this for reference: https://www.evm.codes/#00
Edit: actually that isn't an indication that it should return 0.

andreiburdusa commented:
Yul defines only the type u256, which is the native 256-bit type of the EVM according to https://docs.soliditylang.org/en/latest/yul.html#motivation-and-high-level-description
Therefore the empty bytestring returned by stop would mean 0

andreiburdusa reviewed (approved):
The new changes also look good to me

Coda-Coda committed:
Add info about other primops that are not modelled to README

Coda-Coda committed:
Minor improvement to README
